### PR TITLE
Feature/upgrade spring 159

### PR DESCRIPTION
**Pull request by ltwheeler**
Upgrades Spring to version 1.5.9,

This PR resolves the numerous cyclic dependency injections that were incompatible with the newer version of spring boot.

Issue:
https://trello.com/c/QaLaIrly/87-1164-update-ipa-web-dependency-spring-boot-to-15x

### CODE REVIEW

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java [hunk 1]
```diff
@@ -173,4 +172,169 @@ private String adjustTermCodeYear(String termCode, Long toYear) {
...
 		return newTermCode;
 	}
+
+	/**
+	 * Syncs CRN and location data from DW to IPA, assuming the section/activities already exist
+	 */
+	@Transactional
+	@Override
+	public void updateSectionsFromDW() {
+
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java]: Functions shouldn't begin with blank lines.

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java [hunk 2]
```diff
@@ -173,4 +172,169 @@ private String adjustTermCodeYear(String termCode, Long toYear) {
...
+		List<Course> courses = this.courseService.getAllCourses();
+
+		// Map Keys will look like allDwSections.get(PSC-2017);
+		Map<String, List<DwSection>> allDwSections = new HashMap<>();
+
+		for (Course course : courses) {
+
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java]: for loops shouldn't begin with blank lines (scopes generally shouldn't start with blank lines)

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java [hunk 3]
```diff
@@ -173,4 +172,169 @@ private String adjustTermCodeYear(String termCode, Long toYear) {
...
+			Long year = course.getSchedule().getYear();
+			String subjectCode = course.getSubjectCode();
+			String dwSectionKey = subjectCode + "-" + year;
+
+			// Query the subjectCode/year pair if necessary
+			if (allDwSections.get(dwSectionKey) == null) {
+				List<DwSection> dwSections = dwRepository.getSectionsBySubjectCodeAndYear(subjectCode, year);
+				if (dwSections == null) {
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java]: Combine these two if statements.

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java [hunk 4]
```diff
@@ -173,4 +172,169 @@ private String adjustTermCodeYear(String termCode, Long toYear) {
...
+				}
+
+				allDwSections.put(dwSectionKey, dwSections);
+			}
+
+
+			// Loop through course children
+			for (SectionGroup sectionGroup : course.getSectionGroups()) {
+				for (Section section : sectionGroup.getSections()) {
+
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java]: Remove blank line

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java [hunk 5]
```diff
@@ -173,4 +172,169 @@ private String adjustTermCodeYear(String termCode, Long toYear) {
...
+		this.updateEmptySectionGroups();
+	}
+
+
+	private void updateEmptySectionGroups() {
+		// Find emptySectionGroups in IPA
+		List<SectionGroup> emptySectionGroups = sectionGroupService.findEmpty();
+
+		// Generate termCode+subjectCode pairs to query DW for to get relevant sections
+		List<QueryParam> queryParams = jdbcTemplate.query("SELECT DISTINCT sg.TermCode, c.SubjectCode" +
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java]: Why is jdbcTemplate being used?

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java [hunk 6]
```diff
@@ -173,4 +172,169 @@ private String adjustTermCodeYear(String termCode, Long toYear) {
...
+				}
+			}
+		}
+
+		// Identify activities that exist on all sections, and move them to the sectionGroup sharedActivities
+		for (SectionGroup sectionGroup : emptySectionGroups) {
+			sectionGroupService.identifyAndCondenseSharedActivities(sectionGroup);
+		}
+	}
+
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java]: Remove extra blank line

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java [hunk 7]
```diff
@@ -173,4 +172,169 @@ private String adjustTermCodeYear(String termCode, Long toYear) {
...
+
+	private class QueryParam {
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java]: Rename. QueryParam is too generic of a name.

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaUserRoleService.java [hunk 8]
```diff
@@ -299,5 +300,46 @@ public UserRole findOrAddInstructorRoleToWorkgroup(Workgroup workgroup, User use
...
+	public Instructor findOrAddActiveInstructor(Workgroup workgroup, User user) {
+		Instructor instructor = instructorService.findOrCreate(user.getFirstName(), user.getLastName(), user.getEmail(), user.getLoginId(), workgroup.getId());
+
+		this.findOrAddInstructorRoleToWorkgroup(workgroup, user);
+
+		return instructor;
+	}
+
+	@Override
+	public List<Instructor> findActiveInstructorByScheduleId(long scheduleId) {
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaUserRoleService.java]: Rename method to be plural 'instructors'.

File: src/main/java/edu/ucdavis/dss/ipa/services/InstructorService.java [hunk 9]
```diff
@@ -1,12 +1,8 @@
 package edu.ucdavis.dss.ipa.services;
 
-import edu.ucdavis.dss.ipa.entities.InstructorCost;
-import edu.ucdavis.dss.ipa.entities.User;
-import edu.ucdavis.dss.ipa.entities.Workgroup;
+import edu.ucdavis.dss.ipa.entities.*;
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/InstructorService.java]: Don't use '*' style imports.

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaCourseService.java [hunk 10]
```diff
@@ -19,13 +19,10 @@
 public class JpaCourseService implements CourseService {
 
 	@Inject CourseRepository courseRepository;
+
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaCourseService.java]: Remove blank line

File: src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java [hunk 11]
```diff
@@ -1,26 +1,24 @@
 package edu.ucdavis.dss.ipa.services.jpa;
 
 import javax.inject.Inject;
+import javax.transaction.Transactional;
 
+import edu.ucdavis.dss.dw.dto.DwActivity;
+import edu.ucdavis.dss.dw.dto.DwSection;
 import edu.ucdavis.dss.ipa.entities.*;
+import edu.ucdavis.dss.ipa.services.*;
```
Comment (cthielen) [inline src/main/java/edu/ucdavis/dss/ipa/services/jpa/JpaScheduleOpsService.java]: Remove '*'-style imports

**Comment by ltwheeler** [comment]
Regarding the jdbcTemplate query it looks like we added that code back in July. The query results are custom so there isn't a good pre-existing repo for it to live in. We use jdbcTemplate in a workgroupService method for a similar 'one off' calculation, so it doesn't seem too problematic. Could relocate it to the DwClient maybe. What do you think?

**Comment by cthielen** [comment]
Hm, can you point me to the July commit you're referencing?

**Comment by ltwheeler** [comment]
Here you go,
so the jdbcTemplate query was used in the method updateSectionsFromDW(), which was originally in jpaSectionService, and looks like that code first showed up in this commit:
https://github.com/ucdavis/ipa-web/commit/f7791e5fdb5b5a15bf6a29e9b2b58641adbda3d4#diff-c2d72690ce8a1d80184e941440b1b397
(Ah, from June, not July)

**Pull request merged**